src/main/java/org/mule/extension/email/internal/mailbox/BaseMailboxPollingSource.java

```diff
@@ -0,0 +1,192 @@
+   */
+  @Override
+  protected void doStart() throws ConnectionException {
+    java.util.Optional<? extends BaseEmailPredicateBuilder> builder = getPredicateBuilder();
+    predicate = builder.isPresent() ? builder.get().build() : a -> true;
+    storedEmailContentFactory = new StoredEmailContentFactory();
+    connection = connectionProvider.connect();
+  }
+
+  /**
+   * {@inheritDoc}
+   * <p>
+   * Closes the provided connection.
+   */
+  @Override
+  protected void doStop() {
+    LOGGER.debug("Stopping Email Listener");
+
+    if (connection != null) {
+      connectionProvider.disconnect(connection);
+    }
+  }
+
+  /**
+   * Polls emails from the configured folder.
+   * <p>
+   * This method handles the deletion of the retrieved emails if specified and sets watermark to all the polled items if required.
+   * <p>
+   * The folder is opened at the end of the method to get an updated state of the folder and is closed at the end to save the
+   * state of the mailbox with changes such as flag marks and deletion of the polled elements.
+   * <p>
+   * All retrieved emails are parsed to extract its attributes and content and are ALWAYS opened.
+   */
+  @Override
+  public void poll(PollContext<DefaultStoredEmailContent, BaseEmailAttributes> pollContext) {
+    openFolder = connection.getFolder(folder, READ_WRITE);
+    for (Message message : getMessages(openFolder)) {
+      BaseEmailAttributes attributes = config.parseAttributesFromMessage(message, openFolder);
+      String id = attributes.getId();
+      if (predicate.test(attributes)) {
+        pollContext.accept(item -> {
+          if (isWatermarkEnabled()) {
+            item.setWatermark(Long.valueOf(id));
+          }
+          item.setId(id);
+          item.setResult(Result.<DefaultStoredEmailContent, BaseEmailAttributes>builder()
+              .output(getEmailContent(message, id))
+              .attributes(attributes)
+              .build());
+
+          if (deleteAfterRetrieve) {
+            markAsDeleted(id);
+          }
+        });
+      }
+    }
+    connection.closeFolder(deleteAfterRetrieve);
+  }
+
+  /**
+   * Logs a warning if an email was rejected for processing.
+   */
+  @Override
+  public void onRejectedItem(Result<DefaultStoredEmailContent, BaseEmailAttributes> result,
+                             SourceCallbackContext sourceCallbackContext) {
+    result.getAttributes().ifPresent(a -> LOGGER.warn("Email [" + a.getId() + "] was not processed."));
+  }
+
+  private Message[] getMessages(Folder openFolder) {
+    try {
+      return openFolder.getMessages();
+    } catch (MessagingException e) {
+      throw new EmailListException("Error retrieving emails: " + e.getMessage(), e);
+    }
+  }
+
+  /**
+   * Marks an email as deleted looking it by its UID.
+   */
+  private void markAsDeleted(String id) {
+    try {
+      if (openFolder instanceof POP3Folder) {
+        for (Message message : openFolder.getMessages()) {
```

Comment:
you already had all the messages from the poll, isn't this costly to obtain twice?

Reply:
in any case, use the `getMessages(openFolder)` like before